modules/build/src/main/scala/scala/build/ScalafixArtifacts.scala

```diff
@@ -0,0 +1,157 @@
+            logger,
+            cache.withMessage(s"Downloading scalafix.deps")
+          )
+        )
+
+      ScalafixArtifacts(scalafix.map(_._2), tools.map(_._2))
+    }
+
+  private def fetchOrLoadScalafixProperties(
+    extraRepositories: Seq[Repository],
+    logger: Logger,
+    cache: FileCache[Task]
+  ): Either[BuildException, Properties] =
+    either {
+      val cacheDir  = Directories.directories.cacheDir / "scalafix-props-cache"
+      val cachePath = cacheDir / s"scalafix-interfaces-${Constants.scalafixVersion}.properties"
+
+      val content =
+        if (!os.exists(cachePath)) {
+          val interfacesJar = value(fetchScalafixInterfaces(extraRepositories, logger, cache))
+          val propsData     = value(readScalafixProperties(interfacesJar))
+          if (!os.exists(cacheDir)) os.makeDir(cacheDir)
+          os.write(cachePath, propsData)
+          propsData
+        }
+        else os.read(cachePath)
+      val props  = new Properties()
+      val stream = new ByteArrayInputStream(content.getBytes())
+      props.load(stream)
+      props
+    }
+
+  private def fetchScalafixInterfaces(
+    extraRepositories: Seq[Repository],
+    logger: Logger,
+    cache: FileCache[Task]
+  ): Either[BuildException, Path] =
+    either {
+      val scalafixInterfaces = dep"ch.epfl.scala:scalafix-interfaces:${Constants.scalafixVersion}"
+
+      val fetchResult =
+        value(
+          Artifacts.artifacts(
+            List(scalafixInterfaces).map(Positioned.none),
+            extraRepositories,
+            None,
+            logger,
+            cache.withMessage(s"Downloading scalafix-interfaces ${scalafixInterfaces.version}")
+          )
+        )
+
+      val expectedJarName = s"scalafix-interfaces-${Constants.scalafixVersion}.jar"
+      val interfacesJar = fetchResult.collectFirst {
+        case (_, path) if path.last == expectedJarName => path
+      }
+
+      value(
+        interfacesJar.toRight(new BuildException("Failed to found scalafix-interfaces jar") {})
+      )
+    }
+
+  private def readScalafixProperties(jar: Path): Either[BuildException, String] = {
+    import scala.jdk.CollectionConverters.*
+    val zipFile = new ZipFile(jar.toNIO)
+    val entry = zipFile.getEntries().asScala.find(entry =>
+      entry.getName() == "scalafix-interfaces.properties"
+    )
+    val out =
+      entry.toRight(new BuildException("Failed to found scalafix properties") {})
+        .map { entry =>
+          val stream = zipFile.getInputStream(entry)
+          val bytes  = stream.readAllBytes()
+          new String(bytes)
+        }
+    zipFile.close()
+    out
+  }
+
+  private def scalafixPropsKey(scalaVersion: String): Either[BuildException, String] = {
+    val regex = "(\\d)\\.(\\d+).+".r
+    scalaVersion match {
+      case regex("2", "12")              => Right("scala212")
+      case regex("2", "13")              => Right("scala213")
+      case regex("3", x) if x.toInt <= 3 => Right("scala3LTS")
```

Comment:
> but if we want scala-cli to run future versions of scalafix-cli, that inlining could bring some incompatbilities

Yep, I was thinking about that but decided not to worry for the initial PR as LTS changing won't be happing often.

> I was wondering if it would be worth exposing that logic (currently hidden in the [classloading helper](https://github.com/scalacenter/scalafix/blob/aadace9120b9454dca4e91e319516bd8df30e6aa/scalafix-interfaces/src/main/java/scalafix/interfaces/Scalafix.java#L123-L137) that we need to avoid for scala native) as a simple method in Scalafix.java

Yeah, that might be good idea.

Also, I was thinking maybe it would make sence to make a scala-independent launcher in scalafix project. It would be an exactly the same thing that we have in this PR: takes arguments, selects the scalaVersion for scalafix-cli, fetches using coursier and runs it. Then this logic might be properly shared only in scalafix project.

Reply:
https://github.com/scalacenter/scalafix/issues/2127